rest/doc_api.go

```diff
@@ -266,48 +268,151 @@ func (h *handler) handlePutDoc() error {
 	}()
 
 	docid := h.PathVar("docid")
-	body, err := h.readDocument()
-	if err != nil {
-		return err
-	}
-	if body == nil {
-		return base.ErrEmptyDocument
-	}
 
 	var newRev string
 	var doc *db.Document
 	var ok bool
 
 	roundTrip := h.getBoolQuery("roundtrip")
 
-	if h.getQuery("new_edits") != "false" {
-		// Regular PUT:
+	// If new url parameters are used
+	if h.getQuery("deleted") != "" || h.getQuery("expiry") != "" || h.getQuery("revisions") != "" || h.getQuery("replicator2") != "" {
```

Comment:
It looks like there's relatively little shared code between the 'old' and 'replicator2' ways of doing things.  In the interest of better isolation of this change, can we refactor this to:
if h.getQuery("replicator2") {
   h.handlePutReplicatorDoc()
}
As it's currently written, I think this adds unnecessary complexity into the base handlePutDoc, as well as making it difficult to ensure there aren't any regressions to the base flow in this PR.

Reply:
I agree with splitting this into a separate function, however, I think we should still keep the ability to pass deleted, expiry etc through the query too? Otherwise maybe we should change the ticket as it has deviated a little from that.

Reply:
We should be using those properties if present in the replicator2 path, but I think that the replicator2 query parameter should solely be determining whether we use that flow.

Reply:
Okay. Makes sense. Will get that done then.